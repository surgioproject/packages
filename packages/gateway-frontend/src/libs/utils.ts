import client, { SuccessResponse } from './http'

export const defaultFetcher = <T>(url: string): Promise<T> =>
  client.get<SuccessResponse<T>>(url).then((res) => res.data.data)

export const getDownloadUrl = (
  artifactName: string,
  inline = true,
  accessToken?: string | null,
  artifactParams?: URLSearchParams
): string => {
  const urlObject = new URL(
    `/get-artifact/${artifactName}`,
    window.location.origin
  )
  if (accessToken) {
    urlObject.searchParams.set('access_token', accessToken)
  }
  if (!inline) {
    urlObject.searchParams.set('dl', '1')
  }
  if (artifactParams) {
    for (const [key, value] of artifactParams.entries()) {
      urlObject.searchParams.set(key, value)
    }
  }
  return urlObject.toString()
}

export const getExportProviderUrl = (
  providers: string,
  format: string,
  inline = true,
  accessToken?: string | null
): string => {
  const urlObject = new URL(`/export-providers`, window.location.origin)

  urlObject.searchParams.set('providers', providers)
  urlObject.searchParams.set('format', format)

  if (accessToken) {
    urlObject.searchParams.set('access_token', accessToken)
  }
  if (!inline) {
    urlObject.searchParams.set('dl', '1')
  }
  return urlObject.toString()
}
