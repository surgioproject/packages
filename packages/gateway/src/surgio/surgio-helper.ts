import os from 'os'
import { basename, join } from 'path'
import fs from 'fs-extra'
import { Environment } from 'nunjucks'
import semver from 'semver'
import { Logger } from '@nestjs/common'
import { getEngine } from 'surgio/generator'
import { getProvider, PossibleProviderType } from 'surgio/provider'
import type {
  ArtifactConfig,
  CommandConfig,
  RemoteSnippet,
} from 'surgio/internal'
import { packageJson as corePackageJson, cleanCaches } from 'surgio/internal'
import { TMP_FOLDER_NAME } from 'surgio/constant'
import { createHash } from 'crypto'

export const KEY = 'SURGIO_HELPER'

export class SurgioHelper {
  public remoteSnippetList?: ReadonlyArray<RemoteSnippet>
  public artifactList: ReadonlyArray<ArtifactConfig>
  public providerMap: Map<string, PossibleProviderType> = new Map()
  public readonly templateEngine: Environment
  public readonly configHash: string

  constructor(public cwd: string, public readonly config: CommandConfig) {
    this.artifactList = config.artifacts
    this.templateEngine = getEngine(config.templateDir, {
      clashCore: config.clashConfig?.clashCore,
    })
    this.configHash = this.getConfigSHA256Hash()
  }

  public async init(): Promise<this> {
    await this.checkCoreVersion()
    await this.readProviders()

    return this
  }

  private async readProviders(): Promise<void> {
    const files = await fs.readdir(this.config.providerDir, {
      encoding: 'utf8',
    })

    async function readProvider(
      path: string
    ): Promise<PossibleProviderType | undefined> {
      const provider = await (() => {
        try {
          if (!path.endsWith('.js')) {
            return undefined
          }

          const providerName = basename(path, '.js')

          return getProvider(providerName, require(path))
        } catch (err) {
          Logger.error(`读取 Provider (${path}) 失败: ` + err.message)
          return undefined
        }
      })()

      if (!provider?.type) {
        return undefined
      }

      return provider
    }

    for (const file of files) {
      const result = await readProvider(join(this.config.providerDir, file))
      if (result) {
        this.providerMap.set(result.name, result)
      }
    }
  }

  private async checkCoreVersion(): Promise<void> {
    const gatewayPkgFile = require('../../package.json')
    const peerVersion = gatewayPkgFile.peerDependencies.surgio
    const corePkgVersion = corePackageJson.version as string

    // Pre-release doesn't need to check
    if (corePkgVersion.includes('-')) {
      return
    }

    if (!semver.satisfies(corePkgVersion, peerVersion)) {
      Logger.warn('', undefined, false)
      Logger.warn(
        'Surgio 版本过低，请运行下面命令升级后重新运行！',
        undefined,
        false
      )
      Logger.warn(
        `要求版本 ${peerVersion}，当前版本 ${corePackageJson.version}`,
        undefined,
        false
      )
      Logger.warn('', undefined, false)
      Logger.warn('  命令：', undefined, false)
      Logger.warn('  npm install surgio@latest', undefined, false)
      Logger.warn('', undefined, false)
      throw new Error('Surgio 版本过低')
    }
  }

  private getConfigSHA256Hash(): string {
    return createHash('sha256')
      .update(JSON.stringify(this.config))
      .digest('hex')
  }

  public async cleanCache(): Promise<void> {
    const tmpDir = join(os.tmpdir(), TMP_FOLDER_NAME)

    if (this.remoteSnippetList) {
      Logger.log('已清除远程片段')
      this.remoteSnippetList = undefined
    }

    if (fs.existsSync(tmpDir)) {
      Logger.log('已清除文件缓存')
      await fs.remove(tmpDir)
    }

    await cleanCaches()
    Logger.log('已清除 内存/Redis 缓存')
  }
}
