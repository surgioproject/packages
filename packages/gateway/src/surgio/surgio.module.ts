import { Module, DynamicModule, Global } from '@nestjs/common'
import path from 'path'

import { loadConfig } from 'surgio/build/utils/config'
import { KEY, SurgioHelper } from './surgio-helper'
import { SurgioService } from './surgio.service'

@Global()
@Module({})
export class SurgioModule {
  public static register(options: { readonly cwd: string }): DynamicModule {
    return {
      module: SurgioModule,
      providers: [
        SurgioService,
        {
          provide: KEY,
          useFactory: (): Promise<SurgioHelper> => {
            const configFile = path.join(options.cwd, 'surgio.conf.js')
            const config = loadConfig(options.cwd, configFile)
            const helper = new SurgioHelper(options.cwd, config)

            return helper.init()
          },
        },
      ],
      exports: [SurgioService],
    }
  }
}
