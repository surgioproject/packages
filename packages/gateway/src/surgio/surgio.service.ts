import { Injectable, Inject, HttpException, HttpStatus } from '@nestjs/common'
import { Artifact } from 'surgio/build/generator/artifact'
import { PossibleProviderType } from 'surgio/build/provider'
import { CommandConfig } from 'surgio/build/types'

import { KEY, SurgioHelper } from './surgio-helper'

@Injectable()
export class SurgioService {
  constructor(@Inject(KEY) public surgioHelper: SurgioHelper) {}

  public get config(): CommandConfig {
    return this.surgioHelper.config
  }

  public async getArtifact(
    artifactName: string,
    {
      downloadUrl,
    }: {
      downloadUrl?: string
    } = {}
  ): Promise<Artifact | undefined> {
    const target = this.surgioHelper.artifactList.find(
      (item) => item.name === artifactName
    )

    if (!target) {
      return undefined
    }

    const artifactInstance = new Artifact(
      this.surgioHelper.config,
      {
        ...target,
        downloadUrl,
      },
      {
        remoteSnippetList: this.surgioHelper.remoteSnippetList,
        templateEngine: this.surgioHelper.templateEngine,
      }
    )

    await artifactInstance.init()

    return artifactInstance
  }

  public async exportProvider(
    providerName: string,
    format: string | undefined,
    template: string | undefined,
    options: ExportProviderOptions = {}
  ): Promise<Artifact> {
    const artifactConfig = format
      ? {
          name: `${providerName}.conf`,
          provider: providerName,
          template: undefined,
          templateString: this.getTemplateByFormat(
            format,
            options.filter,
            providerName
          ),
          ...(options.combineProviders
            ? {
                combineProviders: options.combineProviders,
              }
            : null),
        }
      : template
      ? {
          name: `${providerName}.conf`,
          downloadUrl: options.downloadUrl,
          provider: providerName,
          template,
          ...(options.combineProviders
            ? {
                combineProviders: options.combineProviders,
              }
            : null),
        }
      : (() => {
          throw new Error('未指定 format 和 template')
        })()

    const artifactInstance = new Artifact(
      this.surgioHelper.config,
      artifactConfig,
      {
        remoteSnippetList: this.surgioHelper.remoteSnippetList || [],
        templateEngine: this.surgioHelper.templateEngine,
      }
    )

    await artifactInstance.init()

    return artifactInstance
  }

  public async transformArtifact(
    artifactName: string,
    format: string,
    filter?: string
  ): Promise<Artifact | string | undefined> {
    const target = this.surgioHelper.artifactList.find(
      (item) => item.name === artifactName
    )

    if (!target) {
      return undefined
    }

    const artifact = {
      ...target,
      template: undefined,
      templateString: this.getTemplateByFormat(format, filter, target.provider),
    }
    const artifactInstance = new Artifact(this.surgioHelper.config, artifact, {
      remoteSnippetList: this.surgioHelper.remoteSnippetList || [],
      templateEngine: this.surgioHelper.templateEngine,
    })

    await artifactInstance.init()

    return artifactInstance.render()
  }

  public listProviders(): ReadonlyArray<PossibleProviderType> {
    return Array.from(this.surgioHelper.providerMap.values())
  }

  public getTemplateByFormat(
    format: string,
    filter?: string,
    providerName?: string
  ): string {
    switch (format) {
      case 'surge-policy':
        return `{{ getSurgeNodes(nodeList${filter ? `, ${filter}` : ''}) }}`

      case 'qx-server':
        return `{{ getQuantumultXNodes(nodeList${
          filter ? `, ${filter}` : ''
        }) }}`

      case 'clash-provider':
        return [
          '---',
          `{{ {proxies: getClashNodes(nodeList${
            filter ? `, ${filter}` : ''
          })} | yaml }}`,
          '',
        ].join('\n')

      case 'ss':
        return `{{ getShadowsocksNodes(nodeList, ${JSON.stringify(
          providerName || 'Surgio'
        )}) | base64 }}`

      case 'ssr':
        return `{{ getShadowsocksrNodes(nodeList, ${JSON.stringify(
          providerName || 'Surgio'
        )}) | base64 }}`

      case 'v2ray':
        return `{{ getV2rayNNodes(nodeList) | base64 }}`

      default:
        throw new HttpException(
          '参数 format 不存在或不正确',
          HttpStatus.BAD_REQUEST
        )
    }
  }
}

export interface ExportProviderOptions {
  readonly downloadUrl?: string
  readonly filter?: string
  readonly combineProviders?: ReadonlyArray<string>
}
