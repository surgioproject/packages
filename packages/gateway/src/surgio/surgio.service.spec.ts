import { HttpException } from '@nestjs/common'
import { Test, TestingModule } from '@nestjs/testing'
import { join } from 'path'
import { httpClient } from 'surgio/utils'

import { SurgioModule } from './surgio.module'
import { SurgioService } from './surgio.service'

describe('SurgioService', () => {
  let surgioService: SurgioService
  let mockedHttpClient

  beforeEach(async () => {
    jest.clearAllMocks()

    mockedHttpClient = jest.spyOn(httpClient, 'get')

    const module: TestingModule = await Test.createTestingModule({
      providers: [],
      imports: [
        SurgioModule.register({
          cwd: join(__dirname, '../../__tests__/__fixtures__/gateway'),
        }),
      ],
    }).compile()

    surgioService = module.get<SurgioService>(SurgioService)
  })

  it('should be defined', () => {
    expect(surgioService).toBeDefined()
  })

  test('getArtifact should work', async () => {
    const artifact = await surgioService.getArtifact('test.conf')

    expect(artifact).not.toBeUndefined()
    expect(artifact?.render()).toMatchSnapshot()
  })

  test('transformArtifact format should work', async () => {
    expect(
      await surgioService.transformArtifact('test.conf', 'surge-policy')
    ).toMatchSnapshot()
    expect(
      await surgioService.transformArtifact('test.conf', 'qx-server')
    ).toMatchSnapshot()
    expect(
      await surgioService.transformArtifact('test.conf', 'clash-provider')
    ).toMatchSnapshot()
    await expect(
      surgioService.transformArtifact('test.conf', 'unknown')
    ).rejects.toThrowError(HttpException)
  })

  test('transformArtifact filter should work', async () => {
    expect(
      await surgioService.transformArtifact('test.conf', 'surge-policy', {
        filter: 'customFilters.globalFilter',
      })
    ).toMatchSnapshot()
    expect(
      await surgioService.transformArtifact('test.conf', 'qx-server', {
        filter: 'customFilters.globalFilter',
      })
    ).toMatchSnapshot()
    expect(
      await surgioService.transformArtifact('test.conf', 'clash-provider', {
        filter: 'customFilters.globalFilter',
      })
    ).toMatchSnapshot()
  })

  test('listProviders', () => {
    const providers = surgioService.listProviders()

    expect(providers).toHaveLength(6)

    providers.forEach((item) => {
      expect(item.name).toBeDefined()
    })
  })
})
