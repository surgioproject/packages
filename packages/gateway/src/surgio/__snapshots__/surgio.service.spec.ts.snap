// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurgioService getArtifact should work 1`] = `
"ðŸ‡ºðŸ‡¸US 1 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com
ðŸ‡ºðŸ‡¸US 2 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password
ðŸ‡ºðŸ‡¸ US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password
ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
socks = socks5, server, 443
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com
----
ðŸ‡ºðŸ‡¸US 1, ðŸ‡ºðŸ‡¸US 2, ðŸ‡ºðŸ‡¸US 3, ðŸ‡ºðŸ‡¸ US, Snell, HTTPS, ss1, ss2, ss3, vmess, socks, http 1, http 2, snell, ss4, ss-wss
"
`;

exports[`SurgioService transformArtifact filter should work 1`] = `"ðŸ‡ºðŸ‡¸ US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true"`;

exports[`SurgioService transformArtifact filter should work 2`] = `"shadowsocks=us.example.com:443, method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com, udp-relay=true, fast-open=true, tag=ðŸ‡ºðŸ‡¸ US"`;

exports[`SurgioService transformArtifact filter should work 3`] = `
"---
proxies:
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ðŸ‡ºðŸ‡¸ US
    password: password
    port: "443"
    server: us.example.com
    udp: true
    plugin: obfs
    plugin-opts:
      mode: tls
      host: gateway-carry.icloud.com

"
`;

exports[`SurgioService transformArtifact format should work 1`] = `
"ðŸ‡ºðŸ‡¸US 1 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com
ðŸ‡ºðŸ‡¸US 2 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password
ðŸ‡ºðŸ‡¸ US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password
ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
socks = socks5, server, 443
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com"
`;

exports[`SurgioService transformArtifact format should work 2`] = `
"shadowsocks=us.example.com:443, method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com, tag=ðŸ‡ºðŸ‡¸US 1
shadowsocks=us.example.com:443, method=chacha20-ietf-poly1305, password=password, tag=ðŸ‡ºðŸ‡¸US 2
shadowsocks=us.example.com:443, method=chacha20-ietf-poly1305, password=password, obfs=wss, obfs-host=gateway-carry.icloud.com, obfs-uri=/, tls-verification=true, tag=ðŸ‡ºðŸ‡¸US 3
shadowsocks=us.example.com:443, method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com, udp-relay=true, fast-open=true, tag=ðŸ‡ºðŸ‡¸ US
http=us.example.com:443, username=username, password=password, over-tls=true, tls-verification=true, tag=HTTPS
shadowsocks=server:443, method=chacha20-ietf-poly1305, password=password, udp-relay=true, tag=ss1
shadowsocks=server:443, method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=www.bing.com, tag=ss2
shadowsocks=server:443, method=chacha20-ietf-poly1305, password=password, obfs=ws, obfs-host=server, obfs-uri=/, tag=ss3
vmess=server:443, method=chacha20-poly1305, password=uuid, aead=true, tag=vmess
http=server:443, username=username, password=password, over-tls=true, tls-verification=true, tag=http 1
http=server:443, username=username, password=password, tag=http 2
shadowsocks=server:443, method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=example.com, tag=ss4
shadowsocks=server:443, method=chacha20-ietf-poly1305, password=password, obfs=wss, obfs-host=cloudflare.com, obfs-uri=/ws, tls-verification=true, tag=ss-wss"
`;

exports[`SurgioService transformArtifact format should work 3`] = `
"---
proxies:
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ðŸ‡ºðŸ‡¸US 1
    password: password
    port: "443"
    server: us.example.com
    udp: false
    plugin: obfs
    plugin-opts:
      mode: tls
      host: gateway-carry.icloud.com
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ðŸ‡ºðŸ‡¸US 2
    password: password
    port: "443"
    server: us.example.com
    udp: false
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ðŸ‡ºðŸ‡¸US 3
    password: password
    port: "443"
    server: us.example.com
    udp: false
    plugin: v2ray-plugin
    plugin-opts:
      mode: websocket
      tls: true
      host: gateway-carry.icloud.com
      path: /
      mux: false
      headers: {}
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ðŸ‡ºðŸ‡¸ US
    password: password
    port: "443"
    server: us.example.com
    udp: true
    plugin: obfs
    plugin-opts:
      mode: tls
      host: gateway-carry.icloud.com
  - type: snell
    name: Snell
    server: us.example.com
    port: "443"
    psk: password
    obfs-opts:
      mode: tls
  - type: http
    name: HTTPS
    server: us.example.com
    port: "443"
    username: username
    password: password
    tls: true
    skip-cert-verify: false
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ss1
    password: password
    port: 443
    server: server
    udp: true
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ss2
    password: password
    port: 443
    server: server
    udp: false
    plugin: obfs
    plugin-opts:
      mode: tls
      host: www.bing.com
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ss3
    password: password
    port: 443
    server: server
    udp: false
    plugin: v2ray-plugin
    plugin-opts:
      mode: websocket
      tls: false
      host: server
      path: /
      mux: false
      headers: {}
  - type: vmess
    cipher: auto
    name: vmess
    server: server
    port: 443
    udp: false
    uuid: uuid
    network: tcp
    alterId: "32"
  - type: socks5
    name: socks
    server: server
    port: 443
  - type: http
    name: http 1
    server: server
    port: 443
    username: username
    password: password
    tls: true
    skip-cert-verify: false
  - type: http
    name: http 2
    server: server
    port: 443
    username: username
    password: password
  - type: snell
    name: snell
    server: server
    port: 44046
    psk: yourpsk
    obfs-opts:
      mode: http
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ss4
    password: password
    port: 443
    server: server
    udp: false
    plugin: obfs
    plugin-opts:
      mode: tls
      host: example.com
  - type: ss
    cipher: chacha20-ietf-poly1305
    name: ss-wss
    password: password
    port: 443
    server: server
    udp: false
    plugin: v2ray-plugin
    plugin-opts:
      mode: websocket
      tls: true
      skip-cert-verify: false
      host: cloudflare.com
      path: /ws
      mux: false
      headers: {}

"
`;
